crabs/detection_tracking/models.py

```diff
@@ -0,0 +1,157 @@
+    )
+    if coco_model:  # Return the COCO pretrained model for COCO classes.
+        return model
+
+    # Get the number of input features
+    in_features = model.roi_heads.box_predictor.cls_score.in_features
+    # define a new head for the detector with required number of classes
+    model.roi_heads.box_predictor = FastRCNNPredictor(in_features, num_classes)
+
+    return model
+
+
+def train_faster_rcnn(
+    config,
+    model: nn.Module,
+    train_dataloader: DataLoader,
+    optimizer: optim.Optimizer,
+) -> nn.Module:
+    """Train faster rcnn model
+
+    Parameters
+    ----------
+    config
+        config including hyperparameters
+    model: nn.Module
+        the created model for the training
+    train_dataloader: DataLoader
+        dataloader instance for the given dataset which contains images and labels
+    optimizer: optim.Optimizer
+        optimizer instance for the given model
+
+    Returns
+    -------
+    model : nn.Module
+        the trained model
+    """
+
+    # select device (whether GPU or CPU)
+    device = (
+        torch.device("cuda")
+        if torch.cuda.is_available()
+        else torch.device("cpu")
+    )
+
+    # setup tensorboard stuff
+    layout = {
+        "Multi": {
+            "recon_loss": [
+                "Multiline",
+                ["recon_loss/train", "recon_loss/validation"],
+            ],
+            "pred_loss": [
+                "Multiline",
+                ["pred_loss/train", "pred_loss/validation"],
+            ],
+        },
+    }
+    writer = SummaryWriter(f"/tmp/tensorboard/{int(time.time())}")
+    writer.add_custom_scalars(layout)
+
+    def log_scalar(name, value, epoch):
+        """Log a scalar value to both MLflow and TensorBoard"""
+        writer.add_scalar(name, value, epoch)
+        mlflow.log_metric(name, value)
+
+    EXPERIMENT_NAME = "baseline"
+    RUN_NAME = f"run_{datetime.today()}"
+
+    try:
+        EXPERIMENT_ID = mlflow.get_experiment_by_name(
+            EXPERIMENT_NAME
+        ).experiment_id
+    except mlflow.exception.MlflowException:
+        EXPERIMENT_ID = mlflow.create_experiment(EXPERIMENT_NAME)
+
+    with mlflow.start_run(experiment_id=EXPERIMENT_ID, run_name=RUN_NAME):
+        mlflow.log_params(config)
+        mlflow.log_param("n_epoch", config["num_epochs"])
+
+        # # Create a SummaryWriter to write TensorBoard events locally
+        output_dir = tempfile.mkdtemp()
+
+        for epoch in range(config["num_epochs"]):
+            model.train()
+            i = 0
+            for batch_idx, (imgs, annotations) in enumerate(train_dataloader):
+                i += 1
+                imgs = list(img.to(device) for img in imgs)
+                annotations = [
+                    {k: v.to(device) for k, v in t.items()}
+                    for t in annotations
+                ]
+
+                loss_dict = model(imgs, annotations)
+                losses = sum(loss for loss in loss_dict.values())
```

Comment:
I think it's slightly more standard to compute the mean loss of the batch right? (or why the sum?)